### PR TITLE
feat(athena-adapter): support custom retry configurations for boto3 calls

### LINKED ISSUE #480: Materialized Tables - Error: (SlowDown) when calling the DeleteObjects operation
I'm experiencing the following error using the adapter when creating many materialized tables:
```
An error occurred (SlowDown) when calling the DeleteObjects operation (reached max retries: 4): Please reduce your request rate.
```

I think the issues arises on [this line](https://github.com/dbt-athena/dbt-athena/blob/cb7050db8105cd23151ce2f583ee849914d9483b/dbt/include/athena/macros/materializations/models/table/create_table_as.sql#L57), as we do not observe any erroring queries in the Athena logs, indicating that the error occurs earlier in the CTAS statement.

The problem occurs for our setup because we have many materialized tables and do not specify unique locations for each table individually, so we hit the S3 limit which yields a 503 response and we quickly run out of retries. 
I'm a little surprised that we encounter this limit at all, as it is (iirc) ~5k/s and each of the 20-30 tables have <30 parquet files in the prefix. 
The prefixes are also unique in the last part of the prefix as we specify: `s3_data_dir: "s3://my-bucket/temp/tables" s3_data_naming: "schema_table_unique"` in the `profile.yml`.

My proposal is to do something like:

1. Add a new configuration parameter to [AthenaCredentials](https://github.com/dbt-athena/dbt-athena/blob/cb7050db8105cd23151ce2f583ee849914d9483b/dbt/adapters/athena/connections.py#L48C7-L48C24), e.g. `num_s3_retries: int = 4` (to maintain current behaviour).
2. Pass the parameter to `get_boto3_config` e.g. [on line 417](https://github.com/dbt-athena/dbt-athena/blob/cb7050db8105cd23151ce2f583ee849914d9483b/dbt/adapters/athena/impl.py#L417C16-L417C16) and other references.
3. Modify `get_boto3_config` to use this parameter as:
```
def get_boto3_config(num_s3_retries: int) -> Config:
    return Config(
        user_agent_extra="dbt-athena-community/" + importlib.metadata.version("dbt-athena-community"),
        retries={
            'max_attempts': num_s3_retries,
            'mode': 'standard'
        }
    )
```

Does this seem logical? I can open a PR if this would make sense. 

I'd also appreciate any feedback or alternative suggestions!

### PR DESCRIPTION
**Pull request by lukealexmiller**
# Description

This PR introduces a new functionality to the Athena adapter allowing users to specify a custom number of retry attempts for boto3 calls within the dbt-athena-community plugin. The modification adds a parameter to get_boto3_config function, enabling it to accept the number of retries as an argument. This change enhances resilience and configurability by letting users define their retry strategies, accommodating for network inconsistencies or AWS rate limits.
Resolves #480.

### Changes:
- Added a `num_retries` parameter to `get_boto3_config` function to accept custom retry attempts.
- Introduced `num_boto3_retries` as an optional configuration in `AthenaCredentials` to store the number of boto3 retries.
- `get_effective_num_retries` method in `AthenaCredentials` to determine the effective number of retries considering the optional `num_boto3_retries`.
- Altered calls to `get_boto3_config` within `AthenaConnectionManager` and `AthenaAdapter` to pass the effective number of retries.
- Updated unit test `test_get_boto3_config` to validate the retry configuration within the boto3 config.

The default behavior remains unchanged if the `num_boto3_retries` is not set, with the system defaulting to the existing `num_retries` value.

### Testing:
The updated test ensures that the `user_agent_extra` remains correctly set and validates the new retry configuration is applied when `num_boto3_retries` is specified.

### Note:
Ensure your environment configuration aligns with the new changes by providing the `num_boto3_retries` value if custom retry behavior is desired.

## Checklist

- [ ] You followed [contributing section](https://github.com/dbt-athena/dbt-athena#contributing)
- [ ] You kept your Pull Request small and focused on a single feature or bug fix.
- [ ] You added unit testing when necessary
- [ ] You added functional testing when necessary

### CODE REVIEW

File: dbt/adapters/athena/config.py [hunk 1]
```diff
@@ -5,5 +5,8 @@
 
 
 @lru_cache()
-def get_boto3_config() -> config.Config:
-    return config.Config(user_agent_extra="dbt-athena-community/" + importlib.metadata.version("dbt-athena-community"))
+def get_boto3_config(num_retries: int) -> config.Config:
+    return config.Config(
```
Comment (nicor88) [inline dbt/adapters/athena/config.py:9]: don't you want to add a validation that check if `num_retries`  is None?

**Comment by nicor88** [comment]
Overall looks good.
@svdimchenko and @Jrmyy could you have another review?
@lukealexmiller please rebase your branch with the latest main in dbt-athena, thanks.

File: dbt/adapters/athena/connections.py [hunk 2]
```diff
@@ -58,6 +58,7 @@ class AthenaCredentials(Credentials):
     debug_query_state: bool = False
     _ALIASES = {"catalog": "database"}
     num_retries: int = 5
+    num_boto3_retries: Optional[int] = None
```
Comment (svdimchenko) [inline dbt/adapters/athena/connections.py:61]: let's add this param to the README.md file

File: dbt/adapters/athena/connections.py [hunk 3]
```diff
@@ -74,6 +75,10 @@ def type(self) -> str:
     def unique_field(self) -> str:
         return f"athena-{hashlib.md5(self.s3_staging_dir.encode()).hexdigest()}"
 
+    @property
+    def get_effective_num_retries(self) -> int:
+        return self.num_boto3_retries if self.num_boto3_retries is not None else self.num_retries
```
Comment (svdimchenko) [inline dbt/adapters/athena/connections.py:80]:
```suggestion
        return self.num_boto3_retries or self.num_retries
```
we can simplify this a bit if you don't mind

File: dbt/adapters/athena/connections.py [hunk 4]
```diff
@@ -74,6 +75,10 @@ def type(self) -> str:
     def unique_field(self) -> str:
         return f"athena-{hashlib.md5(self.s3_staging_dir.encode()).hexdigest()}"
 
+    @property
+    def get_effective_num_retries(self) -> int:
```
Comment (svdimchenko) [inline dbt/adapters/athena/connections.py:79]:
```suggestion
    def effective_num_retries(self) -> int:
```
as this is a property, let's use corresponding naming

File: dbt/adapters/athena/connections.py [hunk 5]
```diff
@@ -235,7 +240,7 @@ def open(cls, connection: Connection) -> Connection:
                     attempt=creds.num_retries + 1,
                     exceptions=("ThrottlingException", "TooManyRequestsException", "InternalServerException"),
                 ),
-                config=get_boto3_config(),
+                config=get_boto3_config(num_retries=creds.get_effective_num_retries),
```
Comment (svdimchenko) [inline dbt/adapters/athena/connections.py:243]:
```suggestion
                config=get_boto3_config(num_retries=creds.effective_num_retries),
```
same about naming

**Review by svdimchenko — COMMENTED** [review]
@lukealexmiller everything looks nice to me, just let's fix minor comments about naming and docs

**Comment by nicor88** [comment]
@lukealexmiller do you think that you can make the changes that @svdimchenko requested within the day? then we include this in the next release.

**Review by Jrmyy — COMMENTED** [review]
Thanks for the contribution
Except @svdimchenko comments LGTM ✅

**Comment by lukealexmiller** [comment]
Thanks for the feedback all, hopefully this is now ready in time for the next release! btw, most of the doc changes are due to `markdownlint` in the pre-commit check.

**Review by nicor88 — APPROVED** [review]

**Pull request merged**